### PR TITLE
#17422 improved

### PR DESCRIPTION
**Pull request by haampie**
This includes #17422, but fixes a list of issues with it:

- no more strict dependency on cmake 3.5, but simply allow 3 and above (required a patch)
- fixes a problem where `rocclr` was built outside of the staging area and then reused in dependents
- fixes compilation of HsakmtRoct which had missing dependencies and flags that made the build fail
- fixes RPATH issues (required a patch for `hsa-rocr-dev` which is also submitted upstream)
- adds a variant for the amd gpu target for rocfft, because hipcc tries to autodetect the gpu target, for which it needs a binary which it assumes to be in a specific folder which we can't have in spack
- Fixes the tests

### CODE REVIEW

**Comment by haampie** [comment]
I think this is mostly correct now. It mimics the AUR more or less. There are some issues upstream still, such as https://github.com/RadeonOpenCompute/rocminfo/issues/9.

**Comment by srekolam** [comment]
@haampie , We did some additional changes on top of pull request (https://github.com/srekolam/spack/pull/1) . I am thinking we will validate everything  our side including yours and then update PR17422. 
@adamjstewart , I need some inputs on the best way forward.

**Comment by haampie** [comment]
@srekolam I think it's easiest if you would just merge https://github.com/srekolam/spack/pull/1, and then add your changes on top of that, so that we can continue with the review in #17422 and close this PR.

This PR has everything rebased on the spack develop branch, so it's not easy to merge back into your PR. https://github.com/srekolam/spack/pull/1 has all commits cherry-picked and should be easier to work with.

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    root_cmakelists_dir = 'llvm'
+
+    install_targets = ['clang-tidy', 'install']
+
+    def cmake_args(self):
+        args = [
+                '-DCMAKE_VERBOSE_MAKEFILE=1',
+                '-DLLVM_ENABLE_PROJECTS=clang;lld;clang-tools-extra;compiler-rt',
+                '-DLLVM_ENABLE_ASSERTIONS=1'
+               ]
```
Comment (aweits) [inline var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: We're going to need this here as well. tagging @naromero77 to get some extra eyes on this

```
        if self.compiler.name == "gcc":
            gcc_prefix = ancestor(self.compiler.cc, 2)
            args.append("-DGCC_INSTALL_PREFIX=" + gcc_prefix)
```

**Review by aweits — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py (hunk 1 above)
Comment (haampie) [reply var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: Sure, let me add it. What does it do exactly?

Edit: I see, it's also part of the normal llvm package

Comment (naromero77) [reply var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: Yes, that seems necessary on some Linux OS while others don't seem to require it. Safer to just always specify it.

**Comment by haampie** [comment]
@aweits @naromero77: your suggestion `append("-DGCC_INSTALL_PREFIX=" + gcc_prefix)` breaks clang++ for me :(

```
harmen-desktop tmp $ cat main.cc 
#include <iostream>

int main() {
    std::cout << "hello\n";
}

harmen-desktop tmp $ # with the suggestion
harmen-desktop tmp $ /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/llvm-amdgpu-3.5.0-4pxgxhxmrdc6sd7fgz5c5xndmtj6ikqy/bin/clang++ main.cc 
main.cc:1:10: fatal error: 'iostream' file not found
#include <iostream>
         ^~~~~~~~~~
1 error generated.

harmen-desktop tmp $ # without the suggestion
harmen-desktop tmp $ /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/llvm-amdgpu-3.5.0-3mqgueaylvsdnzgenyd4fnqtonhji7vs/bin/clang++ main.cc 
harmen-desktop tmp $ ./a.out 
hello
```

Apparently it gets `GCC_INSTALL_PREFIX=/`.

Edit: this is because `spack compiler find` sets the paths to `/bin/g++` on Ubuntu 20.04. :cry:. I guess it works if I change that to `/usr/bin/g++`

I'm gonna leave it in anyways, see https://github.com/spack/spack/issues/17932, people should just change their system compiler paths then.

**Comment by aweits** [comment]
> @aweits @naromero77: your suggestion `append("-DGCC_INSTALL_PREFIX=" + gcc_prefix)` breaks clang++ for me :(

I'm curious, what does the output from 'clang++ -v' give you in the broken case?

**Comment by naromero77** [comment]
> @aweits @naromero77: your suggestion `append("-DGCC_INSTALL_PREFIX=" + gcc_prefix)` breaks clang++ for me :(
> 
> ```
> harmen-desktop tmp $ cat main.cc 
> #include <iostream>
> 
> int main() {
>     std::cout << "hello\n";
> }
> 
> harmen-desktop tmp $ # with the suggestion
> harmen-desktop tmp $ /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/llvm-amdgpu-3.5.0-4pxgxhxmrdc6sd7fgz5c5xndmtj6ikqy/bin/clang++ main.cc 
> main.cc:1:10: fatal error: 'iostream' file not found
> #include <iostream>
>          ^~~~~~~~~~
> 1 error generated.
> 
> harmen-desktop tmp $ # without the suggestion
> harmen-desktop tmp $ /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/llvm-amdgpu-3.5.0-3mqgueaylvsdnzgenyd4fnqtonhji7vs/bin/clang++ main.cc 
> harmen-desktop tmp $ ./a.out 
> hello
> ```
> 
> Apparently it gets `GCC_INSTALL_PREFIX=/`.
> 
> Edit: this is because `spack compiler find` sets the paths to `/bin/g++` on Ubuntu 20.04. 😢. I guess it works if I change that to `/usr/bin/g++`
> 
> I'm gonna leave it in anyways, see #17932, people should just change their system compiler paths then.

Sure -- sorry, that is annoying

**Comment by adamjstewart** [comment]
Can we close this PR and focus reviews on #17422?

**Comment by haampie** [comment]
I would be in favor of that too, but #17422 is missing some commits (https://github.com/srekolam/spack/pull/2).

**Comment by adamjstewart** [comment]
@srekolam would it be possible to give @haampie merge privileges to your fork so you two can collaborate on a single branch?

**Comment by srekolam** [comment]
@haampie , thanks for your inputs..we have the recipes for rocblas and other math libs too.. we plan to raise the PR once the current 2 are merged into the spack.

**Comment by srekolam** [comment]
@haampie , i see some build failures in HIP after pulling your 2&3 , I have reverted back to commit id 9c6f763814813541633d147fa92497eb11073a68. I will try to cherry pick  others if needed from 2&3

**Comment by srekolam** [comment]
After merging https://github.com/srekolam/spack/pull/2 , i uninstalled everything and rebuilt hsakmt-roct, hsa-rocr-dev,llvm-amdgpu . I then pulled https://github.com/srekolam/spack/pull/3 and tried to build hip.. Not sure if this is causing the issue.. I have actually been able to build and do some tests for https://github.com/srekolam/spack/pull/1..

**Comment by adamjstewart** [comment]
Since all of these changes have been merged into #17422, I'm going to close this so we can focus the review on that PR.

**Pull request closed without merging**